clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```
@@ -1805,15 +1813,23 @@ private boolean processBackgroundEvents() {
      * execution of the rebalancing logic. The rebalancing logic cannot complete until the
      * {@link ConsumerRebalanceListener} callback is performed.
      *
-     * @param future         Event that contains a {@link CompletableFuture}; it is on this future that the
-     *                       application thread will wait for completion
-     * @param timer          Overall timer that bounds how long to wait for the event to complete
+     * @param future                    Event that contains a {@link CompletableFuture}; it is on this future that the
+     *                                  application thread will wait for completion
+     * @param timer                     Overall timer that bounds how long to wait for the event to complete
+     * @param ignoreErrorEventException Predicate to ignore background errors.
+     *                                  Any exceptions found while processing background events that match the predicate won't be propagated.
      * @return {@code true} if the event completed within the timeout, {@code false} otherwise
      */
     // Visible for testing
-    <T> T processBackgroundEvents(Future<T> future, Timer timer) {
+    <T> T processBackgroundEvents(Future<T> future, Timer timer, Predicate<Exception> ignoreErrorEventException) {
         do {
-            boolean hadEvents = processBackgroundEvents();
+            boolean hadEvents = false;
+            try {
+                hadEvents = processBackgroundEvents();
+            } catch (Exception e) {
+                if (!ignoreErrorEventException.test(e))
+                    throw e;
```

Should we set `hadEvents` to true after the exception is swallowed

---

Yes, I will update it within Jira https://issues.apache.org/jira/browse/KAFKA-17647.

---

uhm wouldn't that defeat the purpose of the `hadEvents` and make that we could end up returning from the `processBackgroundEvents` when we find the InvalidTopicsException, but the Unsubscribe hasn't completed yet? 
The purpose of the `hadEvents` is to give the `future` the second chance to complete:

https://github.com/apache/kafka/blob/05696037d330e5e355d62cb1b977bac9df78db82/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1841-L1845

so finding an ErrorEvent to ignore doesn't mean that the Unsubscribe from all topics completed (it just means that at least one of the topics in the subscription was invalid). So I would expect we give the Unsuscribe that second chance to complete if needed, regardless of the error. Makes sense?

---

@lianetm you are right, but the excepted relationship works only between `Unsubscribe` and `ConsumerRebalanceListenerCallbackNeededEvent`, right? I mean how we know what `event` is expected by the `future` and then set the correct value to `hadEvents`? In fact, what I meant to say is that the method `processBackgroundEvents` should have a more descriptive name related to its role in the unsubscribe use case. However, this has already been addressed in #16686 :smile: